Mathlib/Analysis/Normed/Group/SeparationQuotient.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2024 Yoh Tanimoto. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yoh Tanimoto
+-/
+import Mathlib.Analysis.Normed.Group.Uniform
+
+/-!
+# The null subgroup in a seminormed commutative group
+
+For any `SeminormedAddCommGroup M`, a `NormedAddCommGroup` instance has been defined in
+`Mathlib.Analysis.Normed.Group.Uniform`.
+
+## Main definitions
+
+We use `M` to denote seminormed groups.
+All the following definitions are in the `NullSubgroup` namespace. Hence we can access
+`NullSubgroup.normedMk` as `normedMk`.
+
+* `nullSubgroup` : the subgroup of elements `x` with `‖x‖ = 0`.
+
+## Implementation details
+
+For any `SeminormedAddCommGroup M`, we define a norm on `SeparationQuotient M` by
+`‖x‖ = ‖mk x‖` using the lift.
+
+-/
+
+
+noncomputable section
+
+open SeparationQuotient Set
+
+variable {M : Type*} [SeminormedAddCommGroup M]
+
+namespace SeparationQuotient
+
+/-- The null subgroup with respect to the norm. -/
+def nullSubgroup : AddSubgroup M where
+  carrier := {x : M | ‖x‖ = 0}
+  add_mem' {x y} (hx : ‖x‖ = 0) (hy : ‖y‖ = 0) := by
+    apply le_antisymm _ (norm_nonneg _)
+    refine (norm_add_le x y).trans_eq ?_
+    rw [hx, hy, add_zero]
+  zero_mem' := norm_zero
+  neg_mem' {x} (hx : ‖x‖ = 0) := by
+    simp only [mem_setOf_eq, norm_neg]
+    exact hx
+
+@[simp]
+lemma mem_nullSubmodule_iff {x : M} : x ∈ nullSubgroup ↔ ‖x‖ = 0 := Iff.rfl
+
+lemma inseparable_iff_norm_zero (x y : M) : Inseparable x y ↔ ‖x - y‖ = 0 := by
+  rw [Metric.inseparable_iff, dist_eq_norm]
```

Comment:
Is there really no earlier place this can go? I would imagine it can live very near to `dist_eq_norm`.

Reply:
I think this lemma is probably omitted deliberately; you're expected to perform the two rewrites separately, as mathlib has to draw the line against transitive rewrites somewhere.

Reply:
removed the lemma